lib/streamDAO.js

```diff
@@ -53,6 +53,15 @@ function filterDatumForMongo(datum) {
       datum.deviceId = datum.deviceId.replace(BadDeviceIDPrefix, GoodDeviceIDPrefix);
       datum.id = schema.generateId(datum, schema.idFields(datum.type));
     }
+
+    // This is a migration step which converts times to native time objects before write to keep client changes minimal
+    // it would be better if this could be done right at input, but its simpler this way for now.
+    if ('time' in datum && typeof(datum.time) === 'string') {
```

Comment:
This code is called when inserting or updating a datum. I believe deduplication will not work correctly if you first update a datum, then you try to reinsert the same one (with identical identity fields). Update doesn't change `id` and `_id` which are already generated based on the original string value of a date. When inserting a new one, you'll end up generating a new `id` and `_id` based on native date times, which will generate a different hash.